pkg/pool-stable-phantom/contracts/StablePhantomPool.sol

```diff
@@ -244,20 +227,39 @@ contract StablePhantomPool is StablePool {
     }
 
     /**
-     * @dev Initialize phantom BPT
+     * @dev Due to how this pool works, all the BPT needs to be minted initially. On one hand, we cannot do that in the
+     * constructor because the Vault would call back this contract. On the other hand, this pool also requires to be
+     * initialized with a proportional join due to how the Stable math works.
+     * Then, the approach followed is to mint the total amount of BPT to the sender initializing the pool so it can
+     * be fetched by the Vault as part of the initialization process.
      */
     function _onInitializePool(
         bytes32,
+        address sender,
         address,
-        address,
-        uint256[] memory,
-        bytes memory
+        uint256[] memory scalingFactors,
+        bytes memory userData
     ) internal override whenNotPaused returns (uint256, uint256[] memory) {
-        // Mint initial BPTs and adds them to the Vault via a special join
-        _approve(address(this), address(getVault()), _MAX_TOKEN_BALANCE);
-        uint256[] memory amountsIn = new uint256[](_getTotalTokens());
-        amountsIn[_bptIndex] = _MAX_TOKEN_BALANCE;
-        return (_MAX_TOKEN_BALANCE, amountsIn);
+        StablePool.JoinKind kind = userData.joinKind();
+        _require(kind == StablePool.JoinKind.INIT, Errors.UNINITIALIZED);
+
+        uint256[] memory amountsIn = userData.initialAmountsIn();
+        InputHelpers.ensureInputLengthMatch(amountsIn.length, _getTotalTokens());
+        _upscaleArray(amountsIn, scalingFactors);
+
+        (uint256 currentAmp, ) = _getAmplificationParameter();
+        uint256 invariantAfterJoin = StableMath._calculateInvariant(currentAmp, _dropBptItem(amountsIn), true);
+
+        // Set the initial BPT to the value of the invariant
+        uint256 bptAmountOut = invariantAfterJoin;
+        _updateLastInvariant(invariantAfterJoin, currentAmp);
+
+        // Mint the total amount of BPT to the sender forcing the Vault to pull it
+        uint256 initialBpt = _MAX_TOKEN_BALANCE.sub(bptAmountOut);
```

Comment:
Should we have to subtract the minimum bpt also? Because it is going to be minted after calling `_onInitializePool` in the `onJoinPool` hook.

Reply:
Nevermind, I just checked and it has been done in another PR.

Reply:
Exactly